x-pack/plugins/observability_solution/infra/public/pages/logs/shared/page_template.tsx

```diff
@@ -79,7 +79,7 @@ export const LogsPageTemplate: React.FC<LogsPageTemplateProps> = ({
               defaultMessage:
                 'Use the Elastic Agent or Beats to send logs to Elasticsearch. We make it easy with integrations for many popular systems and apps.',
             }),
-            href: basePath + `/app/integrations/browse`,
+            href: onboardingLocator?.useUrl({ category: 'logs' }),
```

Comment:
> [!CAUTION]
> This `useUrl` call is conditionally made (inside a ternary operator), violating the rules of hook.
I can't say why the linter does not complain about that, but you can fix it either by making this call before the ternary operator or switching to the synchronous `onboardingLocator.getRedirectUrl` method.

Reply:
++ indeed I'll remove the optional chaining

Reply:
> This useUrl call is conditionally made (inside a ternary operator), violating the rules of hook.
> I can't say why the linter does not complain about that

It's not a ternary expression, so I assume this is why the linter doesn't complain. I agree that it's not ideal, I have switched to the `getRedirectUrl` function. This is also not ideal, the redirect URL isn't human-readable and it breaks the browser's back functionality (I have to click back twice to get back to the original page), so the string outputted by `useUrl` is a nicer fit.

I actually don't think that this violates the rules of hooks in practice. It's not called conditionally (i.e. it's not in a ternary expression, it's optionally-chained). This is going to call the hook if it is defined, and if not it will not call it because it's not defined. Hooks cannot be called conditionally because they must always run in the same order, for each render.

In any given run, this code should not mess up the order of the hook call chain. The locator would have to somehow become defined on subsequent renders of the component. I believe the only time when the locator would be undefined is if the onboarding plugin were disabled, and in that case the plugin would need to be re-enabled during the same session for its locator to be registered, so it's unlikely this could ever happen, IMO.

It does beg the question how one is expected to use this hook if we are going to say optional chaining is not acceptable. I have found these instances in Kibana where this exact pattern is used:

- https://github.com/elastic/kibana/blob/673a942a22d425c1fc4d034f8984ffb9f8c45fa2/x-pack/plugins/maps/public/inspector/vector_tile_adapter/components/tile_request_tab.tsx#L42
- https://github.com/elastic/kibana/blob/673a942a22d425c1fc4d034f8984ffb9f8c45fa2/x-pack/plugins/ml/public/locator/use_ml_href.ts#L23
- https://github.com/elastic/kibana/blob/673a942a22d425c1fc4d034f8984ffb9f8c45fa2/x-pack/plugins/observability_solution/apm/public/components/shared/license_prompt/index.tsx#L24
- https://github.com/elastic/kibana/blob/673a942a22d425c1fc4d034f8984ffb9f8c45fa2/x-pack/plugins/observability_solution/apm/public/context/license/invalid_license_notification.tsx#L20
- https://github.com/elastic/kibana/blob/673a942a22d425c1fc4d034f8984ffb9f8c45fa2/x-pack/plugins/observability_solution/infra/public/hooks/use_license.ts#L44
- https://github.com/elastic/kibana/blob/673a942a22d425c1fc4d034f8984ffb9f8c45fa2/x-pack/plugins/observability_solution/infra/public/pages/logs/page_content.tsx#L88

There are over half a dozen more cases of this as well.

Another alternative would be to switch the `?` optional chaining operator for the `!` non-null assertion operator. I don't think this is viable though, because there's a very real possibility that the plugin could be disabled for some users and in that instance the'd experience an error.

Reply:
Addressed in https://github.com/elastic/kibana/pull/184477/commits/409a64c2cf124c13a1d5524343c0bc5c603c0cb4.